kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```diff
@@ -0,0 +1,552 @@
+		for j, includeDir := range crate.Source.IncludeDirs {
+			relativized, err := removeProjectRoot(includeDir, projectRoot)
+			if err != nil {
+				fmt.Printf("Error getting relative path between %s and %s, not relativizing:%v\n", includeDir, projectRoot, err)
+				continue
+			}
+
+			project.Crates[i].Source.IncludeDirs[j] = relativized
+		}
+
+		for j, excludeDir := range crate.Source.ExcludeDirs {
+			relativized, err := removeProjectRoot(excludeDir, projectRoot)
+			if err != nil {
+				fmt.Printf("Error getting relative path between %s and %s, not relativizing:%v\n", excludeDir, projectRoot, err)
+				continue
+			}
+			project.Crates[i].Source.ExcludeDirs[j] = relativized
+		}
+
+		relativized, err := removeProjectRoot(crate.RootModule, projectRoot)
+		if err != nil {
+			fmt.Printf("Error getting relative path between %s and %s, not relativizing:%v\n", crate.RootModule, projectRoot, err)
+			continue
+		}
+		project.Crates[i].RootModule = relativized
+
+	}
+}
+
+// Remove the project root from a path, returning the relative path.
+func removeProjectRoot(path string, projectRoot string) (string, error) {
+	newPath, err := filepath.Rel(projectRoot, path)
+	if err != nil {
+		fmt.Printf("Error getting relative path between %s and %s, not relativizing:%v\n", path, projectRoot, err)
+		return "", err
+	}
+	return newPath, nil
+}
+
+// Create a new kzip writer.
+func newKzipWriter(ctx context.Context, outputZipPath string) *kzip.Writer {
+	out, err := vfs.Create(ctx, outputZipPath)
+	if err != nil {
+		log.Fatalf("Error creating output zip: %v\n", err)
+	}
+
+	kzipWriter, err := kzip.NewWriteCloser(out)
+	if err != nil {
+		log.Fatalf("Error creating kzip writer: %v\n", err)
+	}
+	return kzipWriter
+}
+
+// transitiveDepsForCrate performs a breadth-first search on the crate dependency graph
+// starting from the given `crate`. It uses the direct dependency relationships defined
+// in `crateDeps` to find all reachable crates.
+func transitiveDepsForCrate(crate crate, crateDeps map[crateId][]crateId) []crateId {
+	queue := []crateId{crate.CrateId}
+	visited := make(map[crateId]bool)
+	visited[crate.CrateId] = true
+	for len(queue) > 0 {
+		currentCrateId := queue[0]
+		queue = queue[1:] // Dequeue
+
+		// Iterate over the direct dependencies of the current crate
+		for _, dep := range crateDeps[currentCrateId] {
+			// If the dependency hasn't been visited, add it to the queue and mark it as visited
+			if _, ok := visited[dep]; !ok {
+				queue = append(queue, dep)
+				visited[dep] = true
+			}
+		}
+	}
+	transitiveDeps := make([]crateId, 0)
+	for k := range visited {
+		if k != crate.CrateId {
+			transitiveDeps = append(transitiveDeps, k)
+		}
+	}
+	sort.Slice(transitiveDeps, func(i, j int) bool {
+		return transitiveDeps[i] < transitiveDeps[j]
+	})
+
+	return transitiveDeps
+}
+
+// getTransitiveDeps returns a map of all crate IDs to transitive crate dependencies
+func getTransitiveDeps(crates []crate) map[crateId][]crateId {
```

Comment:
the return value could include the crate itself.

This is a reasonable interpretation of "transitive dependencies" and would save duplicate code in `collectCrateSourcesImpl` and `synthesizeProjectJson`